File: call_bridge/membership.py
"""Send and clear org.matrix.msc3401.call.member state events.

The bot sends these events as itself, using a per-guest state key
(``_@bot:server_GUEST_xxx``) so that Element Call sees each guest as a
distinct call participant.
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from mautrix.api import HTTPAPI


async def send_call_member(
    client_api: HTTPAPI,
    room_id: str,
    state_key: str,
    device_id: str,
    session_id: str,
    lk_service_url: str,
    expires_ms: int,
) -> None:
    """Send a call.member state event into *room_id*.

    Args:
        client_api: The bot's mautrix HTTPAPI (``self.client.api``).
        room_id: Matrix room ID to send the event into.
        state_key: Full state key, e.g. ``_@bot:hs_GUEST_abc123``.
        device_id: Synthetic device ID for this guest (e.g. ``GUEST_abc123``).
        session_id: UUID4 session / membership ID.
        lk_service_url: The LiveKit JWT service URL for foci_preferred.
        expires_ms: Membership expiry in milliseconds from now.
    """
    content = {
        "application": "m.call",
        "call_id": "",
        "scope": "m.room",
        "device_id": device_id,
        "expires": expires_ms,
        "created_ts": int(time.time() * 1000),
        "focus_active": {
            "type": "livekit",
            "focus_selection": "oldest_membership",
        },
        "foci_preferred": [
            {
                "type": "livekit",
                "livekit_service_url": lk_service_url,
            }
        ],
    }

    await client_api.request(
        method="PUT",
        path=f"/_matrix/client/v3/rooms/{room_id}/state/org.matrix.msc3401.call.member/{state_key}",
        content=content,
    )


async def clear_call_member(
    client_api: HTTPAPI,
    room_id: str,
    state_key: str,
) -> None:
    """Clear a call.member state event (empty content signals departure).

    Args:
        client_api: The bot's mautrix HTTPAPI.
        room_id: Matrix room ID.
        state_key: The state key to clear.
    """
    await client_api.request(
        method="PUT",
        path=f"/_matrix/client/v3/rooms/{room_id}/state/org.matrix.msc3401.call.member/{state_key}",
        content={},
    )

File: membership.go
package main

import (
	"context"
	"fmt"
	"time"

	"maunium.net/go/mautrix"
	"maunium.net/go/mautrix/event"
	"maunium.net/go/mautrix/id"
)

// callMemberEventType is the Matrix event type for call membership (MSC3401/MSC4143).
var callMemberEventType = event.Type{Type: "org.matrix.msc3401.call.member", Class: event.StateEventType}

// SendCallMember sends a call.member state event into a Matrix room.
//
// The bot sends these events as itself, using a per-guest state key
// (_@bot:server_GUEST_xxx) so that Element Call sees each guest as a
// distinct call participant.
func SendCallMember(
	ctx context.Context,
	client *mautrix.Client,
	roomID id.RoomID,
	stateKey string,
	deviceID string,
	sessionID string,
	lkServiceURL string,
	expiresMS int,
) error {
	content := map[string]interface{}{
		"application": "m.call",
		"call_id":     "",
		"scope":       "m.room",
		"device_id":   deviceID,
		"expires":     expiresMS,
		"created_ts":  time.Now().UnixMilli(),
		"focus_active": map[string]interface{}{
			"type":            "livekit",
			"focus_selection": "oldest_membership",
		},
		"foci_preferred": []map[string]interface{}{
			{
				"type":               "livekit",
				"livekit_service_url": lkServiceURL,
			},
		},
	}

	_, err := client.SendStateEvent(ctx, roomID, callMemberEventType, stateKey, content)
	if err != nil {
		return fmt.Errorf("send call.member: %w", err)
	}
	return nil
}

// ClearCallMember clears a call.member state event (empty content signals departure).
func ClearCallMember(
	ctx context.Context,
	client *mautrix.Client,
	roomID id.RoomID,
	stateKey string,
) error {
	_, err := client.SendStateEvent(ctx, roomID, callMemberEventType, stateKey, map[string]interface{}{})
	if err != nil {
		return fmt.Errorf("clear call.member: %w", err)
	}
	return nil
}
